docs/development/design/fsm.md

```diff
@@ -0,0 +1,81 @@
+# Nexd State Machine
+
+> [Issue #885](https://github.com/nexodus-io/nexodus/issues/885)
+
+## Summary
+
+State management in nexd is becoming more complicated.
+We need a mechanism to allow us to simplify that logic and not rely on crash/restart to retrigger things.
+
+## Proposal
+
+I propose that we implement a Finite State Machine inside of nexd that operates as follows:
+
+```mermaid
+---
+title: Nexd State Machine
+---
+stateDiagram-v2
+    [*] --> UNAUTHENTICATED
+    UNAUTHENTICATED --> AUTHENTICATING: "login"
+    AUTHENTICATING --> AUTHENTICATED: "login_success"
+    AUTHENTICATING --> UNAUTHENTICATED: "login_failed"
+    AUTHENTICATING --> UNAUTHENTICATED: "logout"
+    AUTHENTICATED --> REGISTERING: "register_device"
+    AUTHENTICATED --> UNAUTHENTICATED: "logout"
+    REGISTERING --> REGISTERED: "device_registered"
+    REGISTERING --> UNREGISTERED: "registration_failed"
+    REGISTERED --> UNAUTHENTICATED: "logout"
+    UNREGISTERED --> REGISTERING: "register_device"
+    UNREGISTERED --> UNAUTHENTICATED: "logout"
+    REGISTERED --> UP: "up"
+    REGISTERED --> UNAUTHENTICATED: "logout"
+    REGISTERED --> PEERING: "fetch_peers"
+    PEERING --> UNREGISTERED: "down"
+    PEERING --> UNREGISTERED: "peering_error"
+    PEERING --> UNAUTHENTICATED: "authentication_error"
+    PEERING --> UNAUTHENTICATED: "logout"
+    PEERING --> REGISTERED: "stop"
+    PEERING --> UP: "peered"
+    UP --> UNAUTHENTICATED: "logout"
+    UP --> PEERING: "fetch_peers"
+    UP --> REGISTERTED: "stop"
+```
+
+nexd may be in one of these states:
+
+1. UNAUTHENTICATED: No configuration has been provided
+1. AUTHENTICATING: Performing authentication, either interactively or via token
+1. AUTHENTICATED: User is succesfully authenticated
+1. UNREGISTERED: Authenticated but device has not been registered with the control plane
+1. REGISTERED: Device is registered with the control plane, but data plane is not established
+1. PEERING: Device is connected to control plane and fetching peer listings
+1. UP: Data plane is established
+
+To provide information to the user about the current state, a `nexctl nexd status` command will be updated to include the new states. This can inform the user of the current state and what to do next.
+
+To transition from UNAUTHENTICATED to AUTHENTICATING the "login" transition must occur.
+
+The trigger for this will be a either:
+- A CLI command, for example: `nexctl nexd login`
+- Automatic if credentials were provided
```

Comment:
How about:

- Automatic at startup, either using provided credentials, or by starting the OTP login flow
- Automatic any time `UNAUTHENTICATED` was reached not as the result of an explicit `logout` request from the admin.

This also means we need to store this state somewhere.  Imagine an admin requests `logout`, and then the service restarts for some reason (say, the package got upgraded). We want `nexd` to remember that `logout` was the last requested state.